File: search_in_binary_tree.py
def search_in_binary_tree(tree_array,target):
    class Binary_Tree_Node:
        """This is Binary Tree Node under which all BST functions are defined"""

        def __init__(self,val):
            """Initialize a new Binary_Tree_Node with the given value"""
            self.val = val
            self.left = None
            self.right = None 

        def printer(self):
            """Prints a Node with its left and right value"""
            if self is None:
                print('nothing')
                return
            else:
                print(f'Node is {self.val}')
            if self.left is None:
                print('nothing on left')
            else:
                print(f'Node.left is {self.left.val}')
            if self.right is None:
                print('nothing on right')
            else:
                print(f'Node.right is {self.right.val}')
        
        def insert(self,new_val):
            """Inserts a new Node once root is defined"""
            if new_val < self.val:
                if self.left is None:
                    self.left = Binary_Tree_Node(new_val)
                else:
                    self.left.insert(new_val)
            else:
                if self.right is None:
                    self.right = Binary_Tree_Node(new_val)
                else:
                    self.right.insert(new_val)
        
        def inorder_traversal_target(self,target):
            """Finds the target node through inorder traversal"""
            if self.left:
                x = self.left.inorder_traversal_target(target)
                if x is not None:
                    return x

            if target == self.val:
                return self

            if self.right:
                y = self.right.inorder_traversal_target(target)
                if y is not None:
                    return y

        def array_to_tree(tree_array):
            """Coverts array input to tree"""
            for i in range(len(tree_array)): 
                if i == 0:
                    root = Binary_Tree_Node(tree_array[i])
                else:
                    root.insert(tree_array[i])
            return root
        
    root = Binary_Tree_Node.array_to_tree(tree_array)
    # Uncomment the line below if you want to see the root node with its left and right
    #Binary_Tree_Node.printer(root)

    target_subtree = None
    target_subtree = root.inorder_traversal_target(target)

    # Uncomment the line below if you want to see the target subtree root node with its left and right
    #Binary_Tree_Node.printer(target_subtree)

    return target_subtree

print(search_in_binary_tree( [4,2,7,1,3], 2))

# Time complexity: log(n)
# Space complexity: O(n)

File: search_in_binary_tree.js
function searchInBinaryTree(treeArray,target){

    class BinaryTreeNode {
        
        // constructor definition for each Node
        constructor(val){
            this.val = val;
            this.left = null;
            this.right = null;
        }

        // inserts a new Node once root is defined
        insert(new_val){
            if (new_val < this.val) {
                if (this.left === null) {
                    this.left = new BinaryTreeNode(new_val);
                }
                else{
                    this.left.insert(new_val);
                }
            }
            else{
                if(this.right === null) {
                    this.right = new BinaryTreeNode(new_val);
                }
                else {
                    this.right.insert(new_val);
                }
            }
        }

        // finds the target Node through inorder traversal
        inorderTraversalNode(target){
            if (this.left != null) {
                const left = this.left.inorderTraversalNode(target);
                if (left != null){
                    return left;
                }
            }
            if (target == this.val){
                return this;
            }
            if (this.right != null) {
                const right = this.right.inorderTraversalNode(target);
                if (right != null) {
                    return right;
                }
            }
        }

        // converts input array to BST
        arrayToTree(array){
            const root = new BinaryTreeNode(array[0])
            for (let i=1; i<array.length; i++){
                root.insert(array[i]);
            }
            return root;
        }

    // end of BST class
    }

    // handle function input

    const binaryTree = new BinaryTreeNode();
    const root = binaryTree.arrayToTree(treeArray);

    let targetSubtree = null;
    targetSubtree = root.inorderTraversalNode(target);

    return targetSubtree;
}

console.log(searchInBinaryTree([4,2,7,1,3],2));

// Time complexity: log(n)
// Space complexity: O(n)
